### PR TITLE
Adding performance statistics for AI taks

### PR DESCRIPTION
**Pull request by tibor95**
Now it can count calls of individual tasks and their execution time.
To have it printed one must modify constants in the code

### CODE REVIEW

**Review by hessenfarmer — APPROVED** [review]
code LGTM not tested

**Comment by hessenfarmer** [comment]
@Noordfrees 
this instrumentation code might help us to analyse and improve AI performance. 
Any objections to have it in master as it is completely deactivatable and should not be in any release.

File: src/ai/defaultai.cc [hunk 1]
```diff
@@ -7149,7 +7165,8 @@ uint32_t DefaultAI::msites_built() const {
 void DefaultAI::print_stats(const Time& gametime) {
 
 	if (!kPrintStats) {
-		set_taskpool_task_time(Time(), SchedulerTaskId::kPrintStats);
+		set_taskpool_task_time(
+		   Time(), SchedulerTaskId::kPrintStats);  // NOCOM what does Time() returns??
```
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: Regarding the nocom 
it is returning never to silence this task for the rest of the game 
see
https://github.com/widelands/widelands/blob/ac9152863fddca92f6ced92e455705d55dcd5314/src/base/times.h#L143

**Comment by tibor95** [comment]
I am going to add more stats - avg time and max time, as discussed

File: src/ai/defaultai.cc [hunk 2]
```diff
@@ -354,11 +355,15 @@ void DefaultAI::think() {
 	std::sort(current_task_queue.begin(), current_task_queue.end());
 
 	// Performing tasks from temporary queue one by one
-	for (const SchedulerTask& task : current_task_queue) {
+	for (auto task : current_task_queue) {
```
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: clang-tidy is complaining to make this const

File: src/ai/defaultai.cc [hunk 3]
```diff
@@ -7269,6 +7286,14 @@ void DefaultAI::print_stats(const Time& gametime) {
 	                  persistent_data->ai_personality_mil_upper_limit, msites_in_constr(),
 	                  static_cast<int8_t>(soldier_status_),
 	                  player_statistics.get_modified_player_power(player_number()));
+
+	// 5. printing some performance data
+	log_dbg_time(
+	   gametime, "Player: %d, AI tasks statistics:  call count   ms spent\n", player_number());
+	for (auto task : taskPool) {
```
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: here as well

**Comment by tibor95** [comment]
Statistics implemented, after 30 minutes:

```
[00:40:14.403 game] DEBUG: Player: 2, AI tasks statistics:  call count  ms total   avg     max
[00:40:14.403 game] DEBUG:   productionsites statistics  :     181          7      0      0
[00:40:14.403 game] DEBUG:   print statistics            :       4          3      1      1
[00:40:14.403 game] DEBUG:   check enemy sites           :      81         89      1      3
[00:40:14.403 game] DEBUG:   check militarysites         :     308        887      3     10
[00:40:14.403 game] DEBUG:   flag warehouse Update       :     134         35      0      1
[00:40:14.403 game] DEBUG:   roads check                 :     661        357      1     11
[00:40:14.403 game] DEBUG:   count military vacant       :      84          5      0      0
[00:40:14.403 game] DEBUG:   check buildable fields      :     307      13623     44     85
[00:40:14.403 game] DEBUG:   check unbuildable fields    :     345        597      2     35
```

I will finish other stuff tonight

**Comment by tibor95** [comment]
Ok, should be OK now
BUT, I updated .gitignore, is it OK?

File: src/ai/defaultai.cc (hunk 2 above)
Comment (tibor95) [reply src/ai/defaultai.cc]: Should be fixed. This is beyond my C++ capabilities though :(

**Comment by hessenfarmer** [comment]
> Ok, should be OK now BUT, I updated .gitignore, is it OK?

what was the reason for changing this? 
as this is related to VS @matthiakl any objections?

File: src/ai/defaultai.cc (hunk 2 above)
Comment (hessenfarmer) [reply src/ai/defaultai.cc]: > Should be fixed. This is beyond my C++ capabilities though :(

I don't understand this completely as well however I normally just do what clang-tidy suggests.

**Comment by tibor95** [comment]
In fact, I am using  VS Code and it created such folder. But probably I should just add this, not modify current one

**Comment by hessenfarmer** [comment]
> In fact, I am using VS Code and it created such folder. But probably I should just add this, not modify current one

Yes I think so

**Comment by tibor95** [comment]
gitignore fixed

**Comment by hessenfarmer** [comment]
looks good if the checks are as expected I will merge tomorrow morning. 
Thanks for implementing this. I tested it, but on my machine times are lower by factor 10 at least.  so your machine is much better to check for results.

**Comment by tibor95** [comment]
10x? So I need new machine :)

**Comment by hessenfarmer** [comment]
windows builds failed which might be unrelated to this branch but we need to find out so I merged master to trigger a build.

**Comment by matthiakl** [comment]
> windows builds failed which might be unrelated to this branch but we need to find out so I merged master to trigger a build.

Should be fixed with #5158

**Comment by hessenfarmer** [comment]
found another 2 nits inline, but they should not prevent merging

**Comment by tibor95** [comment]
let me know if any action from me is needed.

**Comment by hessenfarmer** [comment]
> let me know if any action from me is needed.

As we should wait for the other branch to be merged first it would be nice if you could fix the 2 small things I commented meanwhile

**Comment by tibor95** [comment]
Which? The "const" stuff were fixed and pulled yesterday

EDIT: pushed

File: src/ai/defaultai.cc (hunk 2 above)
Comment (tibor95) [reply src/ai/defaultai.cc]: fixed

**Comment by hessenfarmer** [comment]
> Which? The "const" stuff were fixed and pulled yesterday
> 
> EDIT: pushed

maybe the push failed as I can't see it in the web interface . please check. afterwards you might merge master a last time and if the check suite only has g++5 and g++6 release on 18.04 failing feel free to merge

EDIT: I just did the merge of master with the web interface

**Comment by hessenfarmer** [comment]
unfortunately there is a new eror in ubuntu 20, which I can't evaluate.
@Noordfrees  could you have a look at the logs

**Comment by Noordfrees** [comment]
That's just an unrelated transient testsuite error

**Pull request merged**